### ISSUE TITLE
Detect reads of instance attributes before initialization (mypy parity?)

### ISSUE THREAD

**Issue by trag1c** [astral-sh/ty#3910]
### Summary

Opening this as instructed after a brief discussion w/ @AlexWaygood =)

Consider these three:
<table>
<tr>
  <th>foo.py</th>
  <th>bar.py</th>
  <th>baz.py</th>
</tr>
<tr>
  <td>

```py
from typing import final

@final
class Foo:
    def __init__(self, msg: str) -> None:
        print(self.msg)
        self.msg = msg
```
  </td>
  <td>

```py
from typing import final

@final
class Foo:
    def __init__(self) -> None:
        self.set_x()
        print(self.x)

    def set_x(self) -> None:
        self.x = 42
```
  </td>
  <td>

```py
from typing import final

@final
class Foo:
    def __init__(self) -> None:
        print(self.x)

    def set_x(self) -> None:
        self.x = 42
```
  </td>
</tr>
</table>

Here's what ty, mypy, and basedpyright think about these:

<table>
<tr>
  <th>File</th>
  <th><code>ty check --error all</code></th>
  <th><code>mypy --strict</code></th>
  <th><code>basedpyright</code></th>
</tr>
<tr>
  <th>foo.py</th>
  <td><pre>All checks passed!</pre></td>
  <td><pre>foo.py:6: error: Cannot determine type of "msg"  [has-type]
Found 1 error in 1 file (checked 1 source file)</pre></td>
  <td><pre>0 errors, 0 warnings, 0 notes</pre></td>
</tr>
<tr>
  <th>bar.py</th>
  <td><pre>All checks passed!</pre></td>
  <td><pre>Success: no issues found in 1 source file</pre></td>
  <td><pre>0 errors, 0 warnings, 0 notes</pre></td>
</tr>
<tr>
  <th>baz.py</th>
  <td><pre>All checks passed!</pre></td>
  <td><pre>Success: no issues found in 1 source file</pre></td>
  <td><pre>/Users/trag1c/dev/playground/baz.py
  /Users/trag1c/dev/playground/baz.py:9:14 - error: Instance variable "x" is not initialized in the class body or __init__ method (reportUninitializedInstanceVariable)
1 error, 0 warnings, 0 notes</pre></td>
</tr>
</table>

---

Relevant comments from Alex:
> […] that's not so much mypy deliberately detecting the error, more that mypy hit some kind of cycle when trying to infer the type of the attribute and decided to explicitly flag this to the user rather than silently falling back to `Any`
> We have an issue open for us to do similar in these kinds of cyclic cases. Currently we usually silently fall back to `Divergent`, which is basically the same as `Any`
> Assuming that we manage to recover from the cycle without crashing 😆
> […]
> exactly when a type checker hits a cycle, and exactly how it recovers from cycles, is going to vary from type checker to type checker

…which explains the different output for mypy, but it was agreed (brought up by @JelleZijlstra) that it would be useful (and easy to implement!) for ty to detect if an attribute is only assigned in `__init__` (and thus detect reads before assignment). So I guess this is what the issue is really about =)

### Version

ty 0.0.56 (3078f9fa2 2026-07-01)
